> Re: Image-based buffering vs. YOLO-box based buffering - I think the biggest impact we'll see is in videos like the skydiving one, or where the subject is generally walking towards the camera and going from small to big, especially in a Go-Pro video, which might have a super-wide image width. If we base the buffer size off of the image size, for a person that's small in the background, the bounding box might proportionally still be huge, while if it's based of the YOLO box size, it might be sized more to the actual person, which could be better for image tracking.
> 
> Best way to test this would be on an actual GoPro video - but that usually takes some time to track/run through - so I'll look into it when I have more time to let something run
> 
> Re: the YOLO results - Yeah I'll make the change to just have it work with the original results. Wasn't sure if we were using it anywhere so didn't want to mess with it - but if we're not then I'll just modify that.

Never mind on the YOLO results - we do need to make a copy of it. I got this error when I tried to do it without:

```
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\aaron\anaconda3\envs\yolo_bounding_box_test\Lib\multiprocessing\process.py", line 314, in _bootstrap
    self.run()
  File "C:\Users\aaron\anaconda3\envs\yolo_bounding_box_test\Lib\multiprocessing\process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\aaron\Documents\GitHub\freemocap\freemocap\core_processes\process_motion_capture_videos\process_recording_folder.py", line 79, in process_recording_folder
    raise e
  File "C:\Users\aaron\Documents\GitHub\freemocap\freemocap\core_processes\process_motion_capture_videos\process_recording_folder.py", line 69, in process_recording_folder
    image_data_numCams_numFrames_numTrackedPts_XYZ = run_image_tracking_pipeline(
                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\aaron\Documents\GitHub\freemocap\freemocap\core_processes\process_motion_capture_videos\processing_pipeline_functions\image_tracking_pipeline_functions.py", line 47, in run_image_tracking_pipeline
    image_data_numCams_numFrames_numTrackedPts_XYZ = run_image_tracking(
                                                     ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\aaron\Documents\GitHub\freemocap\freemocap\core_processes\process_motion_capture_videos\processing_pipeline_functions\image_tracking_pipeline_functions.py", line 78, in run_image_tracking
    image_data_numCams_numFrames_numTrackedPts_XYZ = process_folder_of_videos(
                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^
message: [2024-03-07T15:09:57.0274] [ERROR] [freemocap.core_processes.process_motion_capture_videos.process_recording_folder] [Process:38980 Thread:37956] ::::: [freemocap.core_processes.process_motion_capture_videos.process_recording_folder:process_recording_folder():76] |||| 2D skeleton detection failed, cannot continue processing.
  File "C:\Users\aaron\Documents\GitHub\skellytracker\skellytracker\process_folder_of_videos.py", line 87, in process_folder_of_videos
    array_list = pool.starmap(process_single_video, tasks)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\aaron\anaconda3\envs\yolo_bounding_box_test\Lib\multiprocessing\pool.py", line 375, in starmap
    return self._map_async(func, iterable, starmapstar, chunksize).get()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\aaron\anaconda3\envs\yolo_bounding_box_test\Lib\multiprocessing\pool.py", line 774, in get
    raise self._value
RuntimeError: Inplace update to inference tensor outside InferenceMode is not allowed.You can make a clone to get a normal tensor before doing inplace update.See https://github.com/pytorch/rfcs/pull/17 for more details.
```

Seems like the `tensor` aspect of the YOLO results doesn't like to be tampered with maybe? It works when making a copy of the results however.

--- a/skellytracker/trackers/yolo_mediapipe_combo_tracker/yolo_mediapipe_combo_tracker.py
+++ b/skellytracker/trackers/yolo_mediapipe_combo_tracker/yolo_mediapipe_combo_tracker.py
@@ -56,7 +56,6 @@ class YOLOMediapipeComboTracker(BaseTracker):
         width_buffer = image.shape[1] * (bounding_box_buffer_percentage / 100.0)
         height_buffer = image.shape[0] * (bounding_box_buffer_percentage / 100.0)
 
-        buffered_yolo_results = copy.deepcopy(yolo_results)
         if box_xyxy.size > 0:
             box_left, box_top, box_right, box_bottom = box_xyxy
 
@@ -75,12 +74,15 @@ class YOLOMediapipeComboTracker(BaseTracker):
                 int(box_left) : int(box_right),
             ]
 
+            buffered_yolo_results = copy.deepcopy(yolo_results)
             buffered_yolo_results[0].boxes.xyxy[0] = torch.tensor([box_left, box_top, box_right, box_bottom])
 
         else:
             # eventually we should not even run mediapipe if no bbox is found
             box_left, box_top, box_right, box_bottom = 0, 0, 0, 0
             cropped_image = image
+
+            buffered_yolo_results = yolo_results
 
         cropped_rgb_image = cv2.cvtColor(cropped_image, cv2.COLOR_BGR2RGB)
 

--- a/skellytracker/trackers/yolo_mediapipe_combo_tracker/yolo_mediapipe_combo_tracker.py
+++ b/skellytracker/trackers/yolo_mediapipe_combo_tracker/yolo_mediapipe_combo_tracker.py
@@ -28,6 +28,7 @@ class YOLOMediapipeComboTracker(BaseTracker):
         min_tracking_confidence=0.5,
         static_image_mode=False,
         smooth_landmarks=True,
+        bounding_box_buffer_percentage=10,
     ):
         super().__init__(
             tracked_object_names=MediapipeModelInfo.mediapipe_tracked_object_names,
@@ -45,16 +46,15 @@ class YOLOMediapipeComboTracker(BaseTracker):
 
         pytorch_model = yolo_object_model_dictionary[model_size]
         self.model = YOLO(pytorch_model)
+        self.bounding_box_buffer_percentage = bounding_box_buffer_percentage
 
     def process_image(self, image: np.ndarray, **kwargs) -> Dict[str, TrackedObject]:
         
         yolo_results = self.model(image, classes=0, max_det=1, verbose=False)
         box_xyxy = np.asarray(yolo_results[0].boxes.xyxy).flatten()
 
-        bounding_box_buffer_percentage = 5 #percent increase in bounding box size 
-        
-        width_buffer = image.shape[1] * (bounding_box_buffer_percentage / 100.0)
-        height_buffer = image.shape[0] * (bounding_box_buffer_percentage / 100.0)
+        width_buffer = image.shape[1] * (self.bounding_box_buffer_percentage / 100.0)
+        height_buffer = image.shape[0] * (self.bounding_box_buffer_percentage / 100.0)
 
         buffered_yolo_results = copy.deepcopy(yolo_results)
         if box_xyxy.size > 0:

